File: core/src/nl/t64/game/rpg/screens/world/entity/GraphicsDoor.java
package nl.t64.game.rpg.screens.world.entity;

import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.utils.Array;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.door.Door;
import nl.t64.game.rpg.constants.Constant;
import nl.t64.game.rpg.screens.world.entity.events.Event;
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent;
import nl.t64.game.rpg.screens.world.entity.events.StateEvent;


public class GraphicsDoor extends GraphicsComponent {

    private final Door door;
    private final Animation<TextureRegion> openAnimation;

    public GraphicsDoor(Door door) {
        this.door = door;
        this.openAnimation = loadOpenAnimation();
    }

    @Override
    public void receive(Event event) {
        if (event instanceof LoadEntityEvent loadEvent) {
            position = loadEvent.position;
            state = loadEvent.state;
        }
        if (event instanceof StateEvent stateEvent) {
            state = stateEvent.state;
        }
    }

    @Override
    public void update(float dt) {
        setFrame(dt);
    }

    @Override
    public void render(Batch batch) {
        batch.draw(currentFrame, position.x, position.y, door.getWidth(), door.getHeight());
    }

    @Override
    public void renderOnMiniMap(Entity entity, Batch batch, ShapeRenderer shapeRenderer) {
        // empty
    }

    @Override
    void setFrame(float dt) {
        if (frameTime >= 1f) {
            frameTime = 1f;
        } else if (state.equals(EntityState.OPENED)) {
            frameTime = frameTime + dt;
        } else {
            frameTime = 0f;
        }
        currentFrame = openAnimation.getKeyFrame(frameTime);
    }

    private Animation<TextureRegion> loadOpenAnimation() {
        TextureRegion[][] textureFrames = Utils.getDoorImage(door.getSpriteId(), (int) door.getWidth());
        Array<TextureRegion> frames = new Array<>(new TextureRegion[]{
                textureFrames[0][0], textureFrames[1][0], textureFrames[2][0], textureFrames[3][0]});
        return new Animation<>(Constant.FAST_FRAMES, frames, Animation.PlayMode.NORMAL);
    }

}

File: core/src/nl/t64/game/rpg/screens/world/entity/GraphicsDoor.kt
package nl.t64.game.rpg.screens.world.entity

import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.utils.Array
import nl.t64.game.rpg.Utils
import nl.t64.game.rpg.components.door.Door
import nl.t64.game.rpg.components.door.DoorType
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.screens.world.entity.events.Event
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent
import nl.t64.game.rpg.screens.world.entity.events.StateEvent


class GraphicsDoor(private val door: Door) : GraphicsComponent() {

    private val openAnimation: Animation<TextureRegion> = loadAnimation(false)
    private val shadowAnimation: Animation<TextureRegion>? =
        if (door.type == DoorType.LARGE_WOODEN_GATE) loadAnimation(true) else null
    private var shadowFrame: TextureRegion? = null

    override fun receive(event: Event) {
        if (event is LoadEntityEvent) {
            position = event.position
            state = event.state!!
        }
        if (event is StateEvent) {
            state = event.state
        }
    }

    override fun update(dt: Float) {
        setFrame(dt)
    }

    override fun render(batch: Batch) {
        batch.draw(currentFrame, position.x, position.y, door.width, door.height)
        if (shadowFrame != null) {
            batch.draw(shadowFrame, position.x, position.y - (door.height / 2f), door.width, door.height)
        }
    }

    override fun renderOnMiniMap(entity: Entity, batch: Batch, shapeRenderer: ShapeRenderer) {
        // empty
    }

    override fun setFrame(dt: Float) {
        frameTime = when {
            frameTime >= 1f -> 1f
            state == EntityState.OPENED -> frameTime + dt
            else -> 0f
        }
        currentFrame = openAnimation.getKeyFrame(frameTime)
        shadowFrame = shadowAnimation?.getKeyFrame(frameTime)
    }

    override fun getAnimation(): Animation<TextureRegion> {
        return openAnimation    // todo, doorshadow is not possible in cutscenes this way atm.
    }

    private fun loadAnimation(isShadow: Boolean): Animation<TextureRegion> {
        val textureFrames = Utils.getDoorImage(door.spriteId, door.width.toInt(), isShadow)
        val frames = Array(arrayOf(textureFrames[0][0], textureFrames[1][0], textureFrames[2][0], textureFrames[3][0]))
        return Animation(Constant.FAST_FRAMES, frames, Animation.PlayMode.NORMAL)
    }

}
